mvc/src/main/java/com/interface21/webmvc/servlet/view/JspView.java

```diff
@@ -11,21 +12,27 @@
 public class JspView implements View {
 
     private static final Logger log = LoggerFactory.getLogger(JspView.class);
-
     public static final String REDIRECT_PREFIX = "redirect:";
 
+    private final String viewName;
+
     public JspView(final String viewName) {
+        this.viewName = viewName;
     }
 
     @Override
     public void render(final Map<String, ?> model, final HttpServletRequest request, final HttpServletResponse response) throws Exception {
-        // todo
+        if (viewName.startsWith(REDIRECT_PREFIX)) {
+            response.sendRedirect(viewName.substring(REDIRECT_PREFIX.length()));
+            return;
+        }
 
         model.keySet().forEach(key -> {
             log.debug("attribute name : {}, value : {}", key, model.get(key));
             request.setAttribute(key, model.get(key));
         });
 
-        // todo
+        RequestDispatcher requestDispatcher = request.getRequestDispatcher(viewName);
+        requestDispatcher.forward(request, response);
```

Comment:
DispatcherServlet의 service 메소드에 있던 위 코드를
JspView의 역할이라고 생각하신 이유가 궁금합니다!

Reply:
``` java
    private void move(final String viewName, final HttpServletRequest request, final HttpServletResponse response) throws Exception {
        // 1번
        if (viewName.startsWith(JspView.REDIRECT_PREFIX)) {
            response.sendRedirect(viewName.substring(JspView.REDIRECT_PREFIX.length()));
            return;
        }
        // 2번
        final var requestDispatcher = request.getRequestDispatcher(viewName);
        requestDispatcher.forward(request, response);
    }
```
저는 2번 부분만 JspView로 이동시켰습니다!
이유는 뷰에 데이터를 전달하는 로직 뒤에 forward가 실행되어야 요청이 최종적으로 렌더링 되기 때문입니다.
forward가 빠지면 요청 속성만 설정하고 끝나기 때문에 요청을 Jsp에게 전달하지 못합니다. 

1번을 DispatcherServlet에 둔 이유는 redirect 특성 때문이었습니다.
redirect가 발생할 시 현재 요청에 대한 페이지가 아닌 리다이렉션 응답에 있는 url에 대한 GET 요청을 수행합니다.
그래서 굳이 사용되지 않을 뷰가 렌더링되기 전에 리다이렉션을 먼저 수행하는게 맞다고 생각했습니다.

참고: [Spring Framework mvc -17.5.3 Redirecting to views](https://docs.spring.io/spring-framework/docs/4.1.x/spring-framework-reference/html/mvc.html#mvc-redirecting)

Reply:
``` java
    private void move(final String viewName, final HttpServletRequest request, final HttpServletResponse response) throws Exception {
        // 1번
        if (viewName.startsWith(JspView.REDIRECT_PREFIX)) {
            response.sendRedirect(viewName.substring(JspView.REDIRECT_PREFIX.length()));
            return;
        }
        // 2번
        final var requestDispatcher = request.getRequestDispatcher(viewName);
        requestDispatcher.forward(request, response);
    }
```
저는 2번 부분만 JspView로 이동시켰습니다!
이유는 뷰에 데이터를 전달하는 로직 뒤에 forward가 실행되어야 요청이 최종적으로 렌더링 되기 때문입니다.
forward가 빠지면 요청 속성만 설정하고 끝나기 때문에 요청을 Jsp에게 전달하지 못합니다. 

1번을 DispatcherServlet에 둔 이유는 redirect 특성 때문이었습니다.
redirect가 발생할 시 현재 요청에 대한 페이지가 아닌 리다이렉션 응답에 있는 url에 대한 GET 요청을 수행합니다.
그래서 굳이 사용되지 않을 뷰가 렌더링되기 전에 리다이렉션을 먼저 수행하는게 맞다고 생각했습니다.

참고: [Spring Framework mvc -17.5.3 Redirecting to views](https://docs.spring.io/spring-framework/docs/4.1.x/spring-framework-reference/html/mvc.html#mvc-redirecting)

Reply:
문서 공유까지~ 너무 감사해요 😄 

> One way to force a redirect as the result of a controller response is for the controller to create and return an instance of Spring’s RedirectView. In this case, DispatcherServlet does not use the normal view resolution mechanism. Rather because it has been given the (redirect) view already, the DispatcherServlet simply instructs the view to do its work.

보내주신 공식 문서를 읽었을 때 이런 문단이 있는데요, 저는 `redirect`와 관련된 요청이더라도 DisptacherServlet이 직접 관여하는게 아닌 View 객체에게 시킨다고 이해가 되네요. 
현재 Spring 객체에서는 `RedirectView`객체에게 위임하는 것으로 보이는데 저희 미션에서는 `JspView`가 `RedirectView`의 역할도 겸한다고 이해해 두 가지 모두 이동시켰습니다.

레모네의 의견도 충분히 설득력있고 타당하다고 느꼈지만 아직까지는 리다이렉트 관련 로직도 JspView에게 가는게 맞다는 생각이 조금 더 큽니다..!
이번 단계에서는 일단 유지하고 리팩토링 진행 시 조금 더 고민해볼게요 🙇